File: bindings/php/tests/MockDaemon.php
<?php

declare(strict_types=1);

namespace Ainos\Tests;

use Ainos\NDJSON;

/**
 * Mock Ainos server daemon for testing.
 *
 * Simulates the Ainos server protocol over TCP, allowing tests
 * to run without a real server. Supports all major API methods
 * and streaming responses.
 *
 * @package Ainos\Tests
 */
class MockDaemon
{
    /** @var string Host to bind to */
    private string $host;

    /** @var int Port to bind to */
    private int $port;

    /** @var string Expected authentication token */
    private string $expectedToken;

    /** @var \Socket|null Server socket */
    private ?\Socket $server = null;

    /** @var \Socket|null Client connection socket */
    private ?\Socket $client = null;

    /** @var bool Whether the daemon is running */
    private bool $running = false;

    /** @var int Number of accepted connections */
    private int $connectionCount = 0;

    /** @var array<string, mixed> Server state */
    private array $state = [
        'models' => [],
        'context' => [],
        'request_count' => 0,
    ];

    /** @var array<string, callable> Custom response handlers */
    private array $handlers = [];

    /** @var array<string, mixed> Predefined responses */
    private array $responses = [];

    /**
     * @param string $host Host to bind to
     * @param int $port Port to bind to
     * @param string $expectedToken Expected auth token
     */
    public function __construct(
        string $host = '127.0.0.1',
        int $port = 0,
        string $expectedToken = 'test-token',
    ) {
        $this->host = $host;
        $this->port = $port;
        $this->expectedToken = $expectedToken;
    }

    /**
     * Destructor - ensures clean shutdown.
     */
    public function __destruct()
    {
        $this->stop();
    }

    /**
     * Start the mock daemon.
     *
     * @return int The port the daemon is listening on
     * @throws \RuntimeException if the daemon fails to start
     */
    public function start(): int
    {
        $this->server = @\socket_create(\AF_INET, \SOCK_STREAM, \SOL_TCP);

        if ($this->server === false) {
            throw new \RuntimeException(
                \sprintf('Failed to create socket: %s', \socket_strerror(\socket_last_error()))
            );
        }

        // Allow address reuse
        @\socket_set_option($this->server, \SOL_SOCKET, \SO_REUSEADDR, 1);

        // Bind to the port (port 0 = auto-assign)
        if (!@\socket_bind($this->server, $this->host, $this->port)) {
            throw new \RuntimeException(
                \sprintf('Failed to bind socket: %s', \socket_strerror(\socket_last_error($this->server)))
            );
        }

        // Get the actual port
        if ($this->port === 0) {
            @\socket_getsockname($this->server, $this->host, $this->port);
        }

        // Listen with a backlog
        if (!@\socket_listen($this->server, 5)) {
            throw new \RuntimeException(
                \sprintf('Failed to listen on socket: %s', \socket_strerror(\socket_last_error($this->server)))
            );
        }

        $this->running = true;

        return $this->port;
    }

    /**
     * Stop the mock daemon.
     */
    public function stop(): void
    {
        $this->running = false;

        try {
            if ($this->client !== null) {
                @\socket_shutdown($this->client, \SHUT_RDWR);
                @\socket_close($this->client);
                $this->client = null;
            }
        } catch (\Throwable) {
            // Ignore
        }

        try {
            if ($this->server !== null) {
                @\socket_shutdown($this->server, \SHUT_RDWR);
                @\socket_close($this->server);
                $this->server = null;
            }
        } catch (\Throwable) {
            // Ignore
        }
    }

    /**
     * Set a custom response handler for a specific method.
     *
     * @param string $method The method name
     * @param callable $handler Function(array $params, array $state): array
     */
    public function setHandler(string $method, callable $handler): void
    {
        $this->handlers[$method] = $handler;
    }

    /**
     * Accept a client connection.
     *
     * @param int $timeoutSeconds Timeout in seconds
     * @return bool True if a connection was accepted
     */
    public function acceptConnection(int $timeoutSeconds = 5): bool
    {
        if (!$this->running || $this->server === null) {
            return false;
        }

        // Set non-blocking for select
        @\socket_set_nonblock($this->server);

        $read = [$this->server];
        $write = null;
        $except = null;

        $result = @\socket_select($read, $write, $except, $timeoutSeconds, 0);

        if ($result === false || $result === 0) {
            return false;
        }

        $this->client = @\socket_accept($this->server);

        if ($this->client === false) {
            return false;
        }

        @\socket_set_block($this->client);
        $this->connectionCount++;

        return true;
    }

    /**
     * Handle a single client request.
     *
     * @return bool True if a request was handled, false if client disconnected
     * @throws \RuntimeException on protocol errors
     */
    public function handleRequest(): bool
    {
        if ($this->client === null) {
            return false;
        }

        // Read a line (NDJSON)
        $line = '';
        $buffer = '';

        while (true) {
            $chunk = @\socket_read($this->client, 1, \PHP_BINARY_READ);

            if ($chunk === false || $chunk === '') {
                return false;
            }

            if ($chunk === "\n") {
                break;
            }

            $buffer .= $chunk;

            // Prevent DoS with large lines
            if (\strlen($buffer) > 65536) {
                throw new \RuntimeException('Request line too long');
            }
        }

        $line = \trim($buffer);

        if ($line === '') {
            return true; // Skip empty lines
        }

        // Parse the request
        $request = \json_decode($line, true);

        if (!\is_array($request)) {
            $this->sendError('', 'Invalid JSON in request');
            return true;
        }

        $method = $request['method'] ?? '';
        $params = $request['params'] ?? [];
        $requestId = $request['id'] ?? '';

        // Verify token
        $token = $params['token'] ?? $request['token'] ?? '';
        if ($token !== $this->expectedToken) {
            $this->sendError($requestId, 'Authentication failed: invalid token', 1011);
            return true;
        }

        // Dispatch to handler
        try {
            $this->handleMethod($method, $params, $requestId);
        } catch (\Throwable $e) {
            $this->sendError($requestId, $e->getMessage(), 0);
        }

        $this->state['request_count']++;

        return true;
    }

    /**
     * Handle a single request and return the response data (for testing without socket).
     *
     * @param array $request The request array
     * @return array The response array
     */
    public function handleDirect(array $request): array
    {
        $method = $request['method'] ?? '';
        $params = $request['params'] ?? [];
        $requestId = $request['id'] ?? 'req-001';

        // Verify token
        $token = $params['token'] ?? $request['token'] ?? '';
        if ($token !== $this->expectedToken) {
            return $this->buildErrorResponse($requestId, 'Authentication failed: invalid token', 1011);
        }

        try {
            return $this->buildResponse($method, $params, $requestId);
        } catch (\Throwable $e) {
            return $this->buildErrorResponse($requestId, $e->getMessage(), 0);
        }
    }

    /**
     * Handle a method call.
     *
     * @param string $method Method name
     * @param array $params Parameters
     * @param string $requestId Request ID
     */
    private function handleMethod(string $method, array $params, string $requestId): void
    {
        $response = $this->buildResponse($method, $params, $requestId);
        $this->sendNDJSON($response);
    }

    /**
     * Build a response for a method.
     *
     * @param string $method Method name
     * @param array $params Parameters
     * @param string $requestId Request ID
     * @return array Response data
     */
    private function buildResponse(string $method, array $params, string $requestId): array
    {
        // Check for custom handler
        if (isset($this->handlers[$method])) {
            return ($this->handlers[$method])($params, $this->state, $requestId);
        }

        // Check for predefined response
        if (isset($this->responses[$method])) {
            $response = $this->responses[$method];
            $response['id'] = $requestId;
            return $response;
        }

        return match ($method) {
            'health' => $this->handleHealth($requestId),
            'status' => $this->handleStatus($requestId),
            'modelList' => $this->handleModelList($requestId),
            'modelLoad' => $this->handleModelLoad($params, $requestId),
            'modelUnload' => $this->handleModelUnload($params, $requestId),
            'infer' => $this->handleInfer($params, $requestId),
            'contextStore' => $this->handleContextStore($params, $requestId),
            'contextRetrieve' => $this->handleContextRetrieve($params, $requestId),
            default => $this->buildErrorResponse($requestId, \sprintf('Unknown method: %s', $method), 0),
        };
    }

    /**
     * Handle health check.
     */
    private function handleHealth(string $requestId): array
    {
        return [
            'id' => $requestId,
            'result' => [
                'status' => 'healthy',
                'uptime' => 3600.0,
                'version' => '1.0.0',
                'memory' => [
                    'used' => 1024 * 1024 * 512,
                    'total' => 1024 * 1024 * 1024 * 8,
                    'percentage' => 6.25,
                ],
                'active_connections' => 1,
                'start_time' => \time() - 3600,
                'checks' => [
                    'database' => 'healthy',
                    'cache' => 'healthy',
                    'models' => 'healthy',
                ],
            ],
        ];
    }

    /**
     * Handle server status.
     */
    private function handleStatus(string $requestId): array
    {
        return [
            'id' => $requestId,
            'result' => [
                'version' => '1.0.0',
                'uptime' => 3600,
                'active_models' => ['gpt-3.5-turbo', 'gpt-4'],
                'total_requests' => $this->state['request_count'],
                'memory' => [
                    'used' => 1024 * 1024 * 1024,
                    'total' => 1024 * 1024 * 1024 * 8,
                    'percentage' => 12.5,
                ],
                'cpu_average' => 45.2,
                'active_connections' => 1,
                'start_time' => \time() - 3600,
                'config' => [
                    'max_connections' => 100,
                    'max_model_size' => '10GB',
                    'log_level' => 'info',
                ],
                'hardware' => [
                    'gpu' => ['NVIDIA A100', 'NVIDIA A100'],
                    'cpu' => 'AMD EPYC 7V12',
                    'memory' => '64GB',
                ],
                'stats' => [
                    'avg_inference_time' => 0.5,
                    'p99_inference_time' => 2.0,
                    'tokens_per_second' => 150.0,
                ],
            ],
        ];
    }

    /**
     * Handle model list.
     */
    private function handleModelList(string $requestId): array
    {
        $models = [
            [
                'name' => 'gpt-3.5-turbo',
                'id' => 'gpt-3.5-turbo',
                'path' => '/models/gpt-3.5-turbo',
                'size' => 1024 * 1024 * 1024 * 4,
                'loaded' => true,
                'metadata' => [
                    'architecture' => 'transformer',
                    'parameters' => '175B',
                    'quantization' => 'fp16',
                ],
                'modified_at' => \time() - 86400,
                'loaded_at' => \time() - 3600,
                'version' => '1.0.0',
                'description' => 'GPT-3.5-Turbo model',
                'license' => 'MIT',
                'author' => 'OpenAI',
            ],
            [
                'name' => 'gpt-4',
                'id' => 'gpt-4',
                'path' => '/models/gpt-4',
                'size' => 1024 * 1024 * 1024 * 8,
                'loaded' => true,
                'metadata' => [
                    'architecture' => 'transformer',
                    'parameters' => '1T',
                    'quantization' => 'fp16',
                ],
                'modified_at' => \time() - 172800,
                'loaded_at' => \time() - 7200,
                'version' => '2.0.0',
                'description' => 'GPT-4 model',
                'license' => 'MIT',
                'author' => 'OpenAI',
            ],
            [
                'name' => 'llama-2-7b',
                'id' => 'llama-2-7b',
                'path' => '/models/llama-2-7b',
                'size' => 1024 * 1024 * 1024 * 4,
                'loaded' => false,
                'metadata' => [
                    'architecture' => 'transformer',
                    'parameters' => '7B',
                    'quantization' => 'q4_0',
                ],
                'modified_at' => \time() - 259200,
                'loaded_at' => null,
                'version' => '1.0.0',
                'description' => 'Llama 2 7B model',
                'license' => 'Llama 2 Community License',
                'author' => 'Meta',
            ],
        ];

        $loadedCount = \count(\array_filter($models, fn($m) => $m['loaded']));

        return [
            'id' => $requestId,
            'result' => [
                'models' => $models,
                'total' => \count($models),
                'loaded_count' => $loadedCount,
                'total_size' => \array_sum(\array_column($models, 'size')),
            ],
        ];
    }

    /**
     * Handle model load.
     */
    private function handleModelLoad(array $params, string $requestId): array
    {
        $modelName = $params['model'] ?? '';

        if ($modelName === '') {
            return $this->buildErrorResponse($requestId, 'Model name is required', 1021);
        }

        $model = [
            'name' => $modelName,
            'id' => $modelName,
            'path' => \sprintf('/models/%s', $modelName),
            'size' => 1024 * 1024 * 1024 * 4,
            'loaded' => true,
            'metadata' => [
                'architecture' => 'transformer',
                'parameters' => 'unknown',
                'quantization' => 'fp16',
            ],
            'modified_at' => \time(),
            'loaded_at' => \time(),
            'version' => '1.0.0',
            'description' => \sprintf('Model %s', $modelName),
            'license' => 'MIT',
            'author' => 'Unknown',
        ];

        // Track loaded model
        $this->state['models'][$modelName] = true;

        return [
            'id' => $requestId,
            'result' => $model,
        ];
    }

    /**
     * Handle model unload.
     */
    private function handleModelUnload(array $params, string $requestId): array
    {
        $modelName = $params['model'] ?? '';

        if ($modelName === '') {
            return $this->buildErrorResponse($requestId, 'Model name is required', 1021);
        }

        if (!isset($this->state['models'][$modelName])) {
            return $this->buildErrorResponse(
                $requestId,
                \sprintf('Model "%s" not found', $modelName),
                1031
            );
        }

        unset($this->state['models'][$modelName]);

        return [
            'id' => $requestId,
            'result' => [
                'success' => true,
                'unloaded' => true,
            ],
        ];
    }

    /**
     * Handle inference (non-streaming).
     */
    private function handleInfer(array $params, string $requestId): array
    {
        $model = $params['model'] ?? '';
        $prompt = $params['prompt'] ?? '';
        $stream = $params['stream'] ?? false;

        if ($model === '') {
            return $this->buildErrorResponse($requestId, 'Model is required', 1021);
        }

        if ($prompt === '') {
            return $this->buildErrorResponse($requestId, 'Prompt is required', 1021);
        }

        // Handle streaming
        if ($stream) {
            $this->handleStreamingInfer($params, $requestId);
            return []; // Response already sent via streaming
        }

        // Generate a mock response
        $responseText = \sprintf(
            'This is a mock response from model "%s" for prompt: %s',
            $model,
            $prompt
        );

        $promptTokens = \str_word_count($prompt) + 10;
        $completionTokens = \str_word_count($responseText) + 5;

        return [
            'id' => $requestId,
            'result' => [
                'id' => $requestId,
                'model' => $model,
                'choices' => [
                    [
                        'index' => 0,
                        'text' => $responseText,
                        'finish_reason' => 'stop',
                        'logprobs' => null,
                    ],
                ],
                'usage' => [
                    'prompt_tokens' => $promptTokens,
                    'completion_tokens' => $completionTokens,
                    'total_tokens' => $promptTokens + $completionTokens,
                    'prompt_time' => 0.05,
                    'completion_time' => 0.15,
                    'total_time' => 0.2,
                    'prompt_tokens_per_second' => (int)($promptTokens / 0.05),
                    'completion_tokens_per_second' => (int)($completionTokens / 0.15),
                ],
                'created' => \time(),
                'object' => 'text_completion',
            ],
        ];
    }

    /**
     * Handle streaming inference.
     */
    private function handleStreamingInfer(array $params, string $requestId): void
    {
        $model = $params['model'] ?? '';
        $prompt = $params['prompt'] ?? '';

        $chunks = [
            ['text' => 'This ', 'finish_reason' => null],
            ['text' => 'is ', 'finish_reason' => null],
            ['text' => 'a ', 'finish_reason' => null],
            ['text' => 'streaming ', 'finish_reason' => null],
            ['text' => 'mock ', 'finish_reason' => null],
            ['text' => 'response.', 'finish_reason' => 'stop'],
        ];

        foreach ($chunks as $index => $chunkData) {
            $chunk = [
                'id' => $requestId,
                'type' => 'stream',
                'result' => [
                    'id' => $requestId,
                    'model' => $model,
                    'text' => $chunkData['text'],
                    'index' => 0,
                    'finish_reason' => $chunkData['finish_reason'],
                    'created' => \time(),
                ],
            ];

            $this->sendNDJSON($chunk);

            // Simulate delay between chunks
            \usleep(10000); // 10ms
        }

        // Send stream end
        $endChunk = [
            'id' => $requestId,
            'type' => 'stream_end',
            'result' => [
                'usage' => [
                    'prompt_tokens' => 10,
                    'completion_tokens' => 20,
                    'total_tokens' => 30,
                ],
            ],
        ];

        $this->sendNDJSON($endChunk);
    }

    /**
     * Handle context store.
     */
    private function handleContextStore(array $params, string $requestId): array
    {
        $key = $params['key'] ?? '';
        $value = $params['value'] ?? null;
        $ttl = (int)($params['ttl'] ?? 3600);

        if ($key === '') {
            return $this->buildErrorResponse($requestId, 'Key is required', 1021);
        }

        $this->state['context'][$key] = [
            'value' => $value,
            'ttl' => $ttl,
            'created_at' => \time(),
            'expires_at' => \time() + $ttl,
        ];

        return [
            'id' => $requestId,
            'result' => [
                'id' => \bin2hex(\random_bytes(8)),
                'key' => $key,
                'value' => $value,
                'ttl' => $ttl,
                'created_at' => \time(),
                'expires_at' => \time() + $ttl,
                'last_accessed_at' => \time(),
                'access_count' => 1,
            ],
        ];
    }

    /**
     * Handle context retrieve.
     */
    private function handleContextRetrieve(array $params, string $requestId): array
    {
        $key = $params['key'] ?? '';

        if ($key === '') {
            return $this->buildErrorResponse($requestId, 'Key is required', 1021);
        }

        if (!isset($this->state['context'][$key])) {
            return [
                'id' => $requestId,
                'result' => null,
            ];
        }

        $entry = $this->state['context'][$key];

        return [
            'id' => $requestId,
            'result' => [
                'id' => \bin2hex(\random_bytes(8)),
                'key' => $key,
                'value' => $entry['value'],
                'ttl' => $entry['ttl'],
                'created_at' => $entry['created_at'],
                'expires_at' => $entry['expires_at'],
                'last_accessed_at' => \time(),
                'access_count' => ($entry['access_count'] ?? 0) + 1,
            ],
        ];
    }

    /**
     * Send an NDJSON-encoded response to the client.
     *
     * @param array $data Response data
     */
    private function sendNDJSON(array $data): void
    {
        if ($this->client === null) {
            return;
        }

        $json = \json_encode($data, \JSON_UNESCAPED_UNICODE | \JSON_UNESCAPED_SLASHES);
        $payload = $json . "\n";

        @\socket_write($this->client, $payload, \strlen($payload));
    }

    /**
     * Send an error response.
     *
     * @param string $requestId Request ID
     * @param string $message Error message
     * @param int $code Error code
     */
    private function sendError(string $requestId, string $message, int $code): void
    {
        $response = $this->buildErrorResponse($requestId, $message, $code);
        $this->sendNDJSON($response);
    }

    /**
     * Build an error response array.
     *
     * @param string $requestId Request ID
     * @param string $message Error message
     * @param int $code Error code
     * @return array
     */
    private function buildErrorResponse(string $requestId, string $message, int $code): array
    {
        return [
            'id' => $requestId,
            'error' => [
                'message' => $message,
                'code' => $code,
            ],
        ];
    }

    /**
     * Run a single request-response cycle for testing.
     * Accepts a connection and handles one request.
     *
     * @param int $timeoutSeconds Timeout for accepting connection
     * @return bool True if a request was handled
     */
    public function runOnce(int $timeoutSeconds = 5): bool
    {
        if (!$this->acceptConnection($timeoutSeconds)) {
            return false;
        }

        return $this->handleRequest();
    }

    /**
     * Get the server port.
     *
     * @return int
     */
    public function getPort(): int
    {
        return $this->port;
    }

    /**
     * Get the server host.
     *
     * @return string
     */
    public function getHost(): string
    {
        return $this->host;
    }

    /**
     * Get the number of accepted connections.
     *
     * @return int
     */
    public function getConnectionCount(): int
    {
        return $this->connectionCount;
    }

    /**
     * Get the server state (for test assertions).
     *
     * @return array
     */
    public function getState(): array
    {
        return $this->state;
    }

    /**
     * Check if the daemon is running.
     *
     * @return bool
     */
    public function isRunning(): bool
    {
        return $this->running;
    }

    /**
     * Set a predefined response for a specific method.
     *
     * @param string $method Method name
     * @param array $response Response data (without 'id')
     * @return void
     */
    public function setResponse(string $method, array $response): void
    {
        $this->responses[$method] = $response;
    }
}

File: bindings/ruby/spec/mock_daemon.rb
# frozen_string_literal: true

require 'socket'
require 'json'
require 'thread'

# A mock Ainos daemon for testing purposes.
#
# Listens on a TCP socket and responds to NDJSON requests with
# configurable responses. Supports testing of normal operations,
# error conditions, and streaming scenarios.
#
# @example
#   daemon = MockDaemon.new
#   daemon.start
#   port = daemon.port
#   # ... run tests against localhost:#{port}
#   daemon.stop
class MockDaemon
  # @return [Integer] the port the daemon is listening on
  attr_reader :port

  # @return [String] the host the daemon is listening on
  attr_reader :host

  # @return [Array<String>] log of received requests
  attr_reader :request_log

  # @return [Boolean] whether the daemon is running
  attr_reader :running

  # Create a new mock daemon.
  #
  # @param host [String] the host to bind to
  # @param port [Integer] the port to bind to (0 for auto-assign)
  # @param auth_token [String, nil] expected auth token
  def initialize(host: '127.0.0.1', port: 0, auth_token: nil)
    @host = host
    @port = port
    @auth_token = auth_token
    @server = nil
    @running = false
    @server_thread = nil
    @request_log = []
    @response_handlers = {}
    @default_handler = nil
    @mutex = Mutex.new
    @client_threads = []
    @delay = 0.0 # Artificial delay for all responses
    @should_fail = false
    @fail_after = nil
    @connection_count = 0
    @stream_scenario = nil
  end

  # Start the mock daemon.
  #
  # @return [self]
  def start
    @mutex.synchronize do
      return self if @running

      @server = TCPServer.new(@host, @port)
      @port = @server.addr[1]
      @running = true
      @connection_count = 0
      @request_log.clear

      @server_thread = Thread.new { accept_connections }
      @server_thread.abort_on_exception = false

      # Wait for the server to be ready
      sleep(0.01) until @server_thread.alive? && @server&.closed? == false
    end

    self
  end

  # Stop the mock daemon.
  #
  # @return [self]
  def stop
    @mutex.synchronize do
      @running = false
      @server&.close
      @server = nil

      @client_threads.each(&:kill)
      @client_threads.clear
    end

    self
  end

  # Register a handler for a specific request type.
  #
  # @param type [String] the request type
  # @yield [request_hash, client_socket] handler
  def on_request(type, &block)
    @response_handlers[type.to_s] = block
  end

  # Set the default handler for unhandled request types.
  #
  # @yield [request_hash, client_socket] handler
  def on_default(&block)
    @default_handler = block
  end

  # Set an artificial delay before responding.
  #
  # @param seconds [Float] the delay in seconds
  def delay=(seconds)
    @delay = seconds.to_f
  end

  # Configure the daemon to fail after a certain number of requests.
  #
  # @param count [Integer, nil] fail after this many requests (nil = never)
  # @param error [String] the error message to return
  def fail_after(count, error: 'Internal server error')
    @fail_after = count
    @fail_error = error
  end

  # Configure a streaming scenario.
  #
  # @param chunks [Array<String>] the text chunks to stream
  # @param finish_reason [String] the finish reason
  def stream_scenario(chunks, finish_reason: 'stop')
    @stream_scenario = {
      chunks: chunks,
      finish_reason: finish_reason
    }
  end

  # Get the number of active connections.
  #
  # @return [Integer] connection count
  def connection_count
    @mutex.synchronize { @connection_count }
  end

  # Wait for a specific number of requests to be received.
  #
  # @param count [Integer] the number of requests to wait for
  # @param timeout [Float] the timeout in seconds
  # @return [Boolean] true if the count was reached
  def wait_for_requests(count, timeout: 5.0)
    start = Time.now
    while @request_log.length < count
      break if Time.now - start >= timeout
      sleep(0.01)
    end
    @request_log.length >= count
  end

  private

  # Accept incoming connections.
  def accept_connections
    loop do
      begin
        client = @server.accept
        @mutex.synchronize { @connection_count += 1 }
        thread = Thread.new(client) { |c| handle_client(c) }
        @client_threads << thread
        @client_threads.reject!(&:alive?)
      rescue IOError
        break unless @running
      rescue StandardError => e
        break unless @running
      end
    end
  end

  # Handle a client connection.
  #
  # @param client [TCPSocket] the client socket
  def handle_client(client)
    log_request = +''

    loop do
      begin
        line = client.gets
        break if line.nil?

        line = line.chomp
        next if line.empty?

        request = JSON.parse(line)
        log_request = line

        @mutex.synchronize do
          @request_log << request.dup
        end

        # Check if we should fail
        if @fail_after && @request_log.length >= @fail_after
          send_error(client, request, @fail_error || 'Simulated failure')
          break
        end

        # Apply artificial delay
        sleep(@delay) if @delay > 0

        # Find and invoke handler
        handler = @response_handlers[request['type']] || @default_handler

        if handler
          handler.call(request, client)
        else
          send_default_response(client, request)
        end
      rescue JSON::ParserError
        send_raw(client, JSON.generate({
          type: 'error', id: 'parse-error', ok: false,
          payload: {}, error: 'Invalid JSON'
        }) + "\n")
      rescue IOError
        break
      rescue StandardError => e
        send_raw(client, JSON.generate({
          type: 'error', id: 'unknown', ok: false,
          payload: {}, error: e.message
        }) + "\n")
        break
      end
    end
  rescue IOError
    # Client disconnected
  ensure
    client.close rescue nil
    @mutex.synchronize { @connection_count -= 1 }
  end

  # Send a default response for a request type.
  #
  # @param client [TCPSocket] the client
  # @param request [Hash] the request
  def send_default_response(client, request)
    case request['type']
    when 'health'
      send_response(client, request, {
        status: 'ok', healthy: true, version: '1.0.0',
        uptime: 3600, active_connections: 1, loaded_models: 2
      })
    when 'status'
      send_response(client, request, {
        version: '1.0.0', started_at: Time.now.iso8601,
        active_connections: 1, total_requests: 100,
        avg_latency_ms: 45.2, loaded_models: 2, total_models: 5,
        memory: { used: '4.2GB', total: '16GB' },
        cpu: { usage: 12.5 }
      })
    when 'ping'
      send_response(client, request, { pong: true })
    when 'model_list'
      send_response(client, request, {
        models: [
          { name: 'llama3', status: 'loaded', version: '1.0',
            size_bytes: 4_700_000_000, capabilities: %w[chat completion] },
          { name: 'mistral', status: 'loaded', version: '2.0',
            size_bytes: 3_200_000_000, capabilities: %w[chat completion code] },
          { name: 'codellama', status: 'unloaded', version: '1.5',
            size_bytes: 5_100_000_000, capabilities: %w[code completion] }
        ]
      })
    when 'model_load'
      send_response(client, request, {
        name: request.dig('payload', 'model') || 'unknown',
        status: 'loaded', loaded_at: Time.now.iso8601
      })
    when 'model_unload'
      send_response(client, request, { unloaded: true })
    when 'infer'
      if request.dig('payload', 'stream')
        send_stream_response(client, request)
      else
        send_response(client, request, {
          text: "This is a mock response to: #{request.dig('payload', 'prompt')}",
          tokens: 42, finish_reason: 'stop',
          model: request.dig('payload', 'model'),
          usage: { prompt_tokens: 10, completion_tokens: 42, total_tokens: 52 },
          timing: { generation_ms: 150 }
        })
      end
    when 'context_store'
      send_response(client, request, {
        created_at: Time.now.iso8601,
        updated_at: Time.now.iso8601,
        token_count: 0
      })
    when 'context_retrieve'
      send_response(client, request, {
        data: { stored: true, original: 'data' },
        created_at: Time.now.iso8601
      })
    when 'auth_check'
      check_auth(client, request)
    else
      send_response(client, request, {})
    end
  end

  # Send a streaming response.
  #
  # @param client [TCPSocket] the client
  # @param request [Hash] the request
  def send_stream_response(client, request)
    if @stream_scenario
      chunks = @stream_scenario[:chunks]
      finish_reason = @stream_scenario[:finish_reason]
    else
      chunks = ['Hello', ', ', 'world', '!']
      finish_reason = 'stop'
    end

    chunks.each_with_index do |text, index|
      stream_chunk = {
        type: 'stream',
        id: request['id'],
        payload: {
          text: text,
          index: index,
          finished: false
        },
        ok: true
      }
      send_raw(client, JSON.generate(stream_chunk) + "\n")
      sleep(0.001) # Small delay to simulate streaming
    end

    # Send end marker
    end_chunk = {
      type: 'stream_end',
      id: request['id'],
      payload: {
        text: '',
        tokens: chunks.join.length / 4,
        finished: true,
        finish_reason: finish_reason
      },
      ok: true
    }
    send_raw(client, JSON.generate(end_chunk) + "\n")
  end

  # Check authentication.
  #
  # @param client [TCPSocket] the client
  # @param request [Hash] the request
  def check_auth(client, request)
    auth = request['auth']
    if @auth_token && auth != "Bearer #{@auth_token}"
      send_raw(client, JSON.generate({
        type: 'auth_result', id: request['id'], ok: false,
        payload: {}, error: 'Invalid token'
      }) + "\n")
    else
      send_response(client, request, { authenticated: true })
    end
  end

  # Send a success response.
  #
  # @param client [TCPSocket] the client
  # @param request [Hash] the request
  # @param payload [Hash] the response payload
  def send_response(client, request, payload)
    response = {
      type: 'result',
      id: request['id'],
      payload: payload,
      ok: true
    }
    send_raw(client, JSON.generate(response) + "\n")
  end

  # Send an error response.
  #
  # @param client [TCPSocket] the client
  # @param request [Hash] the request
  # @param error [String] the error message
  # @param code [String] the error code
  def send_error(client, request, error, code: 'error')
    response = {
      type: 'result',
      id: request['id'],
      payload: { error_code: code },
      ok: false,
      error: error
    }
    send_raw(client, JSON.generate(response) + "\n")
  end

  # Send raw data to the client.
  #
  # @param client [TCPSocket] the client
  # @param data [String] the raw data
  def send_raw(client, data)
    client.write(data)
    client.flush
  end
end

# RSpec shared context for mock daemon tests.
RSpec.shared_context 'with mock daemon' do
  let(:mock_daemon) { MockDaemon.new }
  let(:client) do
    Ainos::Client.new(
      host: '127.0.0.1',
      port: mock_daemon.port,
      token: 'test-token',
      auto_reconnect: false,
      connection_timeout: 2.0,
      read_timeout: 5.0
    )
  end

  before(:each) do
    mock_daemon.start
    client.connect
  end

  after(:each) do
    client.disconnect rescue nil
    mock_daemon.stop rescue nil
  end
end